### DISCUSSION TITLE
Encountering race condition in React from Postgres-change subscription on INSERT+UPDATE operation

### DISCUSSION THREAD

**Discussion by scoover** [supabase/supabase#16104] [Questions]
My React tracks points in a game. When a point finishes, I update the winner of the active point, then insert a new mostly-empty point. The code that performs then UPDATE-then-INSERT operation looks like this...
```
  const [points, setPoints] = useState<point[]>([]);
  const nextPoint = async () => {
    try {
      setIsLoading(true);
      await supabase
        .from("points")
        .update({ ... changes... })
        .eq("id", existing_point.id);
      const { data } = await supabase
        .from("points")
        .insert({ ... new (pending) point data... })
        .select();
      ...
```

I have setup a [Postgres-change subscription](https://supabase.com/docs/guides/realtime/postgres-changes) to listen for changes in other instances of the app— there may be many clients following the game— that looks like this...
```
  useEffect(() => {
    const pointsListener = supabase
      .channel("public:points")
      .on(
        "postgres_changes",
        {
          event: "INSERT",
          schema: "public",
          table: "points",
        },
        (payload) => insertPoint(payload.new as point)
      )
      .on(
        "postgres_changes",
        {
          event: "UPDATE",
          schema: "public",
          table: "points",
        },
        (payload) => updatePoint(payload.new as point)
      )
      .subscribe();

    return () => { supabase.removeChannel(pointsListener); };
  }, []);

  function insertPoint(point: point) {
    setPoints([...points, point]);
  }

  function updatePoint(point: point) {
    setPoints(points.map((p): point => (p.id === point?.id ? { ...p, ...point } : p)));
  }
```

**My issue is that, because the `nextPoint()` UPDATE and INSERT operations are happening so close together, the subscription events that call `updatePoint()` and `insertPoint()` are firing quasi-simultaneously resulting in a `setPoints()` race condition where both changes aren't reliably applied to `points` on all subscribed clients.**

Is there a best way to perform a database change requiring two/multiple operations so that subscriptions see a single or discrete change request? Does [an RPC](https://supabase.com/docs/reference/javascript/rpc) help? Do I need to create a buffer for notifications received by my subscription-change handler? Do I need to guess at an appropriate delay between issuing the update & insert requests in `nextPoint()`?

This seems like it should be a common problem, so I'm surprised that there isn't an obvious solution.
Thanks for your help.

**Comment by spacenate** [comment]
I think if you update one record and insert another, and you have subscribed to both "UPDATE" and "INSERT" events, then you "should" (and do) receive two events 💭 

I believe the primary issue here is a combination of:
* Calling `setState` with a value instead of a function, and
* calculating the new state based on the old state, and
* using asynchronous event handlers

The end result is a stale variable reference -- at some point, `setPoints([...points, point])` gets called from an async function that is bound to an outdated copy of `points`. Here's a minimal demo to help clarify this: https://codesandbox.io/s/vigilant-wave-sj837p?file=/src/App.js

**Reply by scoover** [reply]
I'm certain this race condition with asynchronous data is the cause, @spacenate... and the minimal demo certainly creates a reliable repro case. Thanks for demonstrating that.

_I can't imagine I'm the first person to perform an insert-and-update on an event in my application._ 

The next question is how best to resolve it. 
- Does React provide a better way to track the underlying `points` synchronously? I tried storing `points[]` in a synchronous `useRef` and then applying those updates (asynchronously) to `setPoints()` for use outside the hook. Didn’t help.
- Do I need to add a janky 1/2-second timer between the insert and update operations?
- Do I need to build a buffer for payloads back from my subscriptions?
- Do I need to build command logic that doesn't send a second database change until the first has round-tripped through the subscription handler?
- Does a [Supabase RPC](https://supabase.com/docs/reference/javascript/rpc) call that executes both operations somehow help?
- Do I just need to clumsily fetch-all every time I get an INSERT subscription event?
- Is my database structure inherently flawed?

Thanks again for your thoughts and the example, @spacenate.

**Reply by spacenate** [reply]
IMO the best solution is to pass a function to `useState()` (example: `setPoints(prevState => [...prevState, point])`). This neatly bypasses the stale variable reference issue, assuming you use the `prevState` argument provided (again, the immutable `points` reference may be stale!)

Try it out in the codesandbox and you'll see what I mean. If you use the functional form with `setValues()`, you should see the "correct" number of items added to the array when you click multiple times quickly (updates are no longer dropped! 🕺) , but notice the suffix does not increment smoothly (since it comes from a "potentially" stale reference to `values.length`).

Another solution (basically the same thing with a different API) is using `useReducer` instead of `useState`. State changes are emitted as stateless "actions" (example: `dispatch("AddPointAction", point)`), that are applied to your component state via the reducer function you specify.
